apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfSubscriber.cs

```diff
@@ -0,0 +1,723 @@
+            private readonly CancellationTokenSource _softStopCts;
+            private readonly CancellationTokenSource _hardStopCts;
+
+            private readonly object _qLock = new object();
+            private readonly Queue<string> _ackQueue = new Queue<string>();
+            private readonly Queue<string> _extendQueue = new Queue<string>();
+            private int _unAckedMsgCount;
+            private TaskCompletionSource<int> _qEvent = new TaskCompletionSource<int>();
+
+            // States:
+            // Push faulted: local immediate stop/restart streaming-pull request
+            // - push task faults, which cancels streamingPullCts
+            // Pull faulted: local immediate stop/restart streaming-pull request
+            // - pull task faults, which cancels streamingPullCts
+            // soft-stop requested: no more msg processing/receiving, stop pull; continue push, then local hard-stop when all acks done
+            // - Pull task completes, which does not cancel streamingPullCts
+            // hard-stop requested: immediate stop of pull and push
+            // - Pull and Push fault, if they are still running
+
+            public async Task StartAsync()
+            {
+                // Loop to restart pull if it fails recoverably.
+                while (!_hardStopCts.IsCancellationRequested)
+                {
+                    SubscriberClient.StreamingPullStream pull = null;
+                    try
+                    {
+                        var streamingPullCts = CancellationTokenSource.CreateLinkedTokenSource(_hardStopCts.Token);
+                        pull = _client.StreamingPull(CallSettings.FromCancellationToken(streamingPullCts.Token),
+                            new BidirectionalStreamingSettings(10));// TODO: Put back to 1 once bug #188 released in v2.1.0
+                        // Initial call to start subscribe messages arriving.
+                        await _taskHelper.ConfigureAwait(pull.WriteAsync(new StreamingPullRequest
+                        {
+                            SubscriptionAsSubscriptionName = _subscriptionName,
+                            StreamAckDeadlineSeconds = _modifyDeadlineSeconds
+                        }));
+                        // Start Push task which pushes acks and extends.
+                        var pushTask = _taskHelper.Run(() => Push(pull, streamingPullCts.Token));
+                        // Read incoming messages whilst not stopped, and not in a fault condition.
+                        var pullTask = _taskHelper.Run(() => Pull(pull, streamingPullCts.Token));
+                        // Wait for either pull or push to complete.
+                        var completedTask = await _taskHelper.ConfigureAwait(_taskHelper.WhenAny(pushTask, pullTask));
+                        // Only cancel streaming if one of the tasks faulted. Otherwise use the non-abortive shutdown.
+                        if (completedTask.IsFaulted)
```

Comment:
What if one finishes normally and the other finished with an exception, both at pretty much the same time? There's a race condition as to which we see as completedTask here. Does that matter?